File: python/src/loop_engineering_mcp/verification_runner.py
"""Execute verification commands with timeout and error handling."""

import asyncio
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Optional


@dataclass
class VerificationResult:
    """Result of running a verification command."""

    success: bool
    exit_code: int
    stdout: str
    stderr: str
    timed_out: bool = False
    duration_seconds: float = 0.0
    error: Optional[str] = None


class VerificationRunner:
    """Runs shell verification commands in a subprocess."""

    def __init__(self, workspace_root: Path, default_timeout: int = 600):
        self.workspace_root = workspace_root
        self.default_timeout = default_timeout

    async def run(
        self,
        command: str,
        *,
        timeout: Optional[int] = None,
        cwd: Optional[Path] = None,
    ) -> VerificationResult:
        """Run a verification command and capture output."""
        if not command or command.strip() == "echo 'No verification configured'":
            return VerificationResult(
                success=True,
                exit_code=0,
                stdout="No verification configured — skipped",
                stderr="",
            )

        work_dir = cwd or self.workspace_root
        timeout_secs = timeout or self.default_timeout
        start = asyncio.get_event_loop().time()

        try:
            proc = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=str(work_dir),
                env=os.environ.copy(),
            )
            try:
                stdout_bytes, stderr_bytes = await asyncio.wait_for(
                    proc.communicate(), timeout=timeout_secs
                )
            except asyncio.TimeoutError:
                proc.kill()
                await proc.wait()
                duration = asyncio.get_event_loop().time() - start
                return VerificationResult(
                    success=False,
                    exit_code=-1,
                    stdout="",
                    stderr=f"Verification timed out after {timeout_secs}s",
                    timed_out=True,
                    duration_seconds=duration,
                    error="timeout",
                )

            duration = asyncio.get_event_loop().time() - start
            stdout = stdout_bytes.decode("utf-8", errors="replace")
            stderr = stderr_bytes.decode("utf-8", errors="replace")
            exit_code = proc.returncode or 0

            return VerificationResult(
                success=exit_code == 0,
                exit_code=exit_code,
                stdout=stdout,
                stderr=stderr,
                duration_seconds=duration,
            )
        except Exception as e:
            duration = asyncio.get_event_loop().time() - start
            return VerificationResult(
                success=False,
                exit_code=-1,
                stdout="",
                stderr=str(e),
                duration_seconds=duration,
                error=str(e),
            )

File: typescript/src/verification-runner.ts
/**
 * Execute verification commands with timeout and error handling
 */

import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface VerificationResult {
  success: boolean;
  exitCode: number;
  stdout: string;
  stderr: string;
  timedOut: boolean;
  durationSeconds: number;
  error?: string;
}

export class VerificationRunner {
  private workspaceRoot: string;
  private defaultTimeout: number;

  constructor(workspaceRoot: string, defaultTimeout = 600) {
    this.workspaceRoot = workspaceRoot;
    this.defaultTimeout = defaultTimeout;
  }

  async run(
    command: string,
    options: { timeout?: number; cwd?: string } = {}
  ): Promise<VerificationResult> {
    if (!command || command.trim() === "echo 'No verification configured'") {
      return {
        success: true,
        exitCode: 0,
        stdout: 'No verification configured — skipped',
        stderr: '',
        timedOut: false,
        durationSeconds: 0,
      };
    }

    const cwd = options.cwd ?? this.workspaceRoot;
    const timeoutMs = (options.timeout ?? this.defaultTimeout) * 1000;
    const start = Date.now();

    try {
      const { stdout, stderr } = await execAsync(command, {
        cwd,
        timeout: timeoutMs,
        maxBuffer: 10 * 1024 * 1024,
        shell: process.platform === 'win32' ? 'cmd.exe' : '/bin/sh',
      });
      const duration = (Date.now() - start) / 1000;
      return {
        success: true,
        exitCode: 0,
        stdout,
        stderr,
        timedOut: false,
        durationSeconds: duration,
      };
    } catch (error: unknown) {
      const duration = (Date.now() - start) / 1000;
      const execError = error as { code?: number; killed?: boolean; stdout?: string; stderr?: string; message?: string };
      if (execError.killed) {
        return {
          success: false,
          exitCode: -1,
          stdout: execError.stdout ?? '',
          stderr: `Verification timed out after ${timeoutMs / 1000}s`,
          timedOut: true,
          durationSeconds: duration,
          error: 'timeout',
        };
      }
      return {
        success: false,
        exitCode: execError.code ?? -1,
        stdout: execError.stdout ?? '',
        stderr: execError.stderr ?? execError.message ?? String(error),
        timedOut: false,
        durationSeconds: duration,
        error: execError.message,
      };
    }
  }
}
